### PR TITLE
Property-based tests for the base64 module

### PR DESCRIPTION
**Pull request by Maria-12648430**
As the title says, this PR introduces property-based tests for the `base64` module.

Test coverage is around ~95%, looks like some obscure edge cases that may occur in relation to `mime_decode` are rarely created by the generators. However, those are covered by the normal CT suite.

<hr />

Some things I noticed while working on this:
* The decoding functions may fail in a variety of ways on malformed input. There is `badarg` which is fine and expected, but there is also `badarith`, `function_clause`, `case_clause` etc bubbling up from internal functions. This multitude of error reasons makes handling them tedious (and so, lazy as I am, I settled on catching `error:_` after hitting three different error reasons). What (I guess) I'm trying to say is, they essentially all boil down to the argument being somehow bad, and so it may be good to unite them all under the `badarg` reason.
* `mime_decode/1` and `mime_decode_to_string/1`, the relaxed siblings of the pedantic `decode/1` and `decode_to_string/1` functions, are surprisingly strict when it comes to padding, or rather _missing_ padding. In reality (that is informally, which in turn is IMO), padding is not strictly _required_ if you know that your input _is_ Base64. So it may be opportune to lift that restriction for the relaxed decoding functions.

### CODE REVIEW

File: lib/stdlib/test/base64_property_test_SUITE.erl [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%% See the License for the specific language governing permissions and
+%% limitations under the License.
+%%
+%% %CopyrightEnd%
+%%
+-module(base64_property_test_SUITE).
+
+-include_lib("common_test/include/ct.hrl").
+-compile(export_all).
```
Comment (bjorng) [inline lib/stdlib/test/base64_property_test_SUITE.erl:23]:
```suggestion
-compile([export_all, nowarn_export_all]).
```

File: lib/stdlib/test/base64_property_test_SUITE.erl [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+    decode_malformed_case,
+    decode_noisy_case,
+    decode_to_string_case,
+    decode_to_string_malformed_case,
+    decode_to_string_noisy_case,
+    mime_decode_case,
+    mime_decode_malformed_case,
+    mime_decode_to_string_case,
+    mime_decode_to_string_malformed_case
+].
```
Comment (bjorng) [inline lib/stdlib/test/base64_property_test_SUITE.erl:38]:
```suggestion
all() ->
    [
     encode_case,
     encode_to_string_case,
     decode_case,
     decode_malformed_case,
     decode_noisy_case,
     decode_to_string_case,
     decode_to_string_malformed_case,
     decode_to_string_noisy_case,
     mime_decode_case,
     mime_decode_malformed_case,
     mime_decode_to_string_case,
     mime_decode_to_string_malformed_case
    ].
```

File: lib/stdlib/test/property_test/base64_prop.erl [hunk 3]
```diff
@@ -0,0 +1,429 @@
...
+%% distributed under the License is distributed on an "AS IS" BASIS,
+%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%% See the License for the specific language governing permissions and
+%% limitations under the License.
+%%
+%% %CopyrightEnd%
+%%
+-module(base64_prop).
+
+-compile(export_all).
```
Comment (bjorng) [inline lib/stdlib/test/property_test/base64_prop.erl:22]:
```suggestion
-compile([export_all, nowarn_export_all]).
```

File: lib/stdlib/test/property_test/base64_prop.erl [hunk 4]
```diff
@@ -0,0 +1,429 @@
...
+        lists:seq(16#21, 16#2A) ++
+        % +
+        lists:seq(16#2C, 16#2E) ++
+        % /, 0-9
+        lists:seq(16#3A, 16#40) ++
+        % A-Z
+        lists:seq(16#5B, 16#60) ++
+        % a-z
+        lists:seq(16#7B, 16#FF)
+    ).
```
Comment (bjorng) [inline lib/stdlib/test/property_test/base64_prop.erl:255]: It is hard to understand whether this code produces the correct list of non-base64 characters.

I suggest defining a helper function called `b64_chars/0` (which can be used also by the `b64_char/0` generator).

```suggestion
    oneof(lists:seq(16#00, 16#ff) -- b64_chars()).
```

**Review by bjorng — CHANGES_REQUESTED** [review]
Overall looks good to me.

The `base64` module has been optimised for speed (but it is still considered [too slow](https://github.com/erlang/otp/issues/5639)) and no attempt has been made to produce nice error messages. With improvements in the compiler in recent releases, it might now be possible to produce more consistent error behaviour without compromising performance, but that should be a the subject of a separate pull request.

File: lib/stdlib/test/property_test/base64_prop.erl [hunk 5]
```diff
@@ -0,0 +1,429 @@
...
+    b64_equals(list_to_binary(L), B);
+b64_equals(B, L) when is_list(L) ->
+    b64_equals(B, list_to_binary(L));
+b64_equals(B1, B2) when is_binary(B1), is_binary(B2) ->
+    b64_equals1(B1, B2).
+
+b64_equals1(<<Eq:4/bytes>>, <<Eq:4/bytes>>) ->
+    is_b64_binary(Eq);
+b64_equals1(<<Eq:4/bytes, More1/binary>>, <<Eq:4/bytes, More2/binary>>) ->
+    case lists:all(fun is_b64_char/1, binary_to_list(Eq))) of
```
[the remark this answers is not in the document]
Comment (Maria-12648430) [reply lib/stdlib/test/property_test/base64_prop.erl:362]: Now how on earth did _that_ happen? :thinking: I'm pretty sure that I ran it just before pushing... And what's more, why didn't it cause a failure anywhere in the GitHub CI, instead it says "All checks have passed"?

Anyway, I'll fix it :wink:

File: lib/stdlib/test/property_test/base64_prop.erl [hunk 6]
```diff
@@ -0,0 +1,429 @@
...
+        true ->
+            b64_equals1(More1, More2);
+        false ->
+            false
+    end;
+b64_equals1(<<Eq, B1, $=, $=>>, <<Eq, B2, $=, $=>>) ->
+    %% If the encoded string ends with "==", there exist multiple
+    %% possibilies for the character preceding the "==" as only the
```
[the remark this answers is not in the document]
Comment (Maria-12648430) [reply lib/stdlib/test/property_test/base64_prop.erl:370]: :woman_facepalming:

File: lib/stdlib/test/property_test/base64_prop.erl [hunk 7]
```diff
@@ -0,0 +1,429 @@
...
+    end.
+
+%% Check if the given argument is a base64 string
+%% (see is_b64_binary/1)
+is_b64_string(S) ->
+    is_b64_binary(list_to_binary(S)).
+
+%% Check if the argument is a list of bytes.
+is_bytelist(L) ->
+    lists:all(fun (B) -> B >= 16#00 andalso B =< 16#FF end, L).
```
[the remark this answers is not in the document]
Comment (Maria-12648430) [reply lib/stdlib/test/property_test/base64_prop.erl:338]: Ah, right... I left it open to floats.

**Comment by Maria-12648430** [comment]
@bjorng ok, I think I did all the requested changes :smiley:

File: lib/stdlib/test/property_test/base64_prop.erl (hunk 5 above)
[the remark this answers is not in the document]
Comment (bjorng) [reply lib/stdlib/test/property_test/base64_prop.erl]: When the tests are run in Github actions, `proper` is not installed so all property-based test cases will be skipped.

**Comment by bjorng** [comment]
Thanks! Can you squash the commits?

We will test this PR in our daily builds after the release of Erlang/OTP 25.

**Review by bjorng — APPROVED** [review]

**Comment by Maria-12648430** [comment]
> Thanks! Can you squash the commits?

Done :+1:

> We will test this PR in our daily builds after the release of Erlang/OTP 25.

Ok, looking forward to it (OTP/25, too :wink:)

File: lib/stdlib/test/property_test/base64_prop.erl (hunk 5 above)
[the remark this answers is not in the document]
Comment (Maria-12648430) [reply lib/stdlib/test/property_test/base64_prop.erl]: I see.

**Comment by bjorng** [comment]
Thanks for your pull request.

File: lib/stdlib/test/property_test/base64_prop.erl [hunk 8]
```diff
@@ -0,0 +1,429 @@
...
+    list_to_binary(Noise);
+insert_noise([B|Bs] = B64, [N|Ns] = Noise) ->
+    case rand:uniform(2) of
+        1 ->
+            [B|insert_noise(Bs, Noise)];
+        2 ->
+            [N|insert_noise(B64, Ns)]
+    end;
+insert_noise(<<B, Bs/binary>> = B64, [N|Ns] = Noise) ->
+    case rand:uniform(2) of
```
Comment (seriyps) [inline lib/stdlib/test/property_test/base64_prop.erl:283]: it is usually a bad idea to use any side-effects in PBT generators, because indeterminism would completely mess shrinking. If you need some source of entropy, it's better to outsource it to the PBT framework (in this case, by using smth like `list(integer(1, 2))`), so the randomness of the generator is fully controlled by the framework.
